gitoxide-core/src/repository/blame.rs

```diff
@@ -18,40 +18,31 @@ pub fn blame_file(mut repo: gix::Repository, file: &OsStr, out: impl std::io::Wr
         .into();
     let file_path: &BStr = gix::path::os_str_into_bstr(file)?;
 
-    let blame_entries = gix::blame::file(
+    let outcome = gix::blame::file(
         &repo.objects,
         traverse,
         &mut resource_cache,
         work_dir.clone(),
         file_path,
     )?;
-
-    let absolute_path = work_dir.join(file);
-    let file_content = std::fs::read_to_string(absolute_path)?;
-    let lines = file_content.lines();
-
-    write_blame_entries(out, lines, blame_entries)?;
+    write_blame_entries(out, outcome)?;
 
     Ok(())
 }
 
-fn write_blame_entries(
-    mut out: impl std::io::Write,
-    mut lines: Lines<'_>,
-    blame_entries: Vec<gix::blame::BlameEntry>,
-) -> Result<(), std::io::Error> {
-    for blame_entry in blame_entries {
-        for line_number in blame_entry.range_in_blamed_file {
-            let line = lines.next().unwrap();
-
-            writeln!(
+fn write_blame_entries(mut out: impl std::io::Write, outcome: gix::blame::Outcome) -> Result<(), std::io::Error> {
+    for (entry, lines_in_hunk) in outcome.entries_with_lines() {
+        for ((actual_lno, source_lno), line) in entry
+            .range_in_blamed_file
+            .zip(entry.range_in_original_file)
+            .zip(lines_in_hunk)
+        {
+            write!(
                 out,
-                "{} {} {}",
-                blame_entry.commit_id.to_hex_with_len(8),
-                // `line_number` is 0-based, but we want to show 1-based line numbers (as `git`
-                // does).
-                line_number + 1,
-                line
+                "{short_id} {line_no} {src_line_no} {line}",
+                line_no = actual_lno + 1,
+                src_line_no = source_lno + 1,
+                short_id = entry.commit_id.to_hex_with_len(8),
             )?;
         }
     }
```

Comment:
This does not show the correct lines in the final output. See, for example, `cargo run -- blame STABILITY.md` where it seems to use `src_line_no` where it should use `line_no` instead.

Reply:
Funnily enough just now, before your message, I also noticed that the "unify how lines…" commit messes up printing quite a bit. Whether line-number are inverted I don't know, but the lines themselves seem to be strange.

```
7d8d1678 228 230     Ok(gix_discover::repository::Path::from_dot_git_dir(
7bed2a96 229 229
53adfe1c 230 230     Ok(gix_discover::repository::Path::from_dot_git_dir(
7bed2a96 231 231         dot_git,
7bed2a96 232 232         if bare {
20f962e5 233 233             gix_discover::repository::Kind::PossiblyBare
7bed2a96 234 234         } else {
53adfe1c 235 235             gix_discover::repository::Kind::WorkTree { linked_git_dir: None }
7bed2a96 236 236         },
7d8d1678 237 239     .expect("by now the `dot_git` dir is valid as we have accessed it"))
7bed2a96 238 238     )
7bed2a96 239 239     .expect("by now the `dot_git` dir is valid as we have accessed it"))
7bed2a96 240 240 }
7bed2a96 241 241
0ec2389e 242 244 }
0ec2389e 243 245
7bed2a96 244 244 }
7bed2a96 245 245
7bed2a96 246 246 fn bool(v: bool) -> &'static str {
7bed2a96 247 247     match v {
7bed2a96 248 248         true => "true",
7bed2a96 249 249         false => "false",
7bed2a96 250 250     }
7bed2a96 251 251 }

gitoxide ( gix-blame) [$] took 7s
❯ sgs
+ 0 gix-blame   # Add initial implementation and tests for `gix-blame`.
+ 1 gix-api     # feat: Add `blame` plumbing crate to the top-level.
+ 2 gix-cli     # feat: add `gix blame` to the CLI
+ 3 review      # first review round
+ 4 structure   # modularlize `gix-blame/lib.rs`
> 5 unify-lines # unify how lines in blame results are accessed
- 6 slimdown    # remove duplication and unnecessary parameter
- 7 add-stats   # feat: Add support for statistics and additional performance information.
- 8 error       # add `Error` type
- 9 todos       # Replace todos!() with... something else

gitoxide ( gix-blame) [$]
❯ cargo run --bin gix -- --verbose --trace blame gix/src/create.rs
```

```
7d8d1678 228     };
7bed2a96 229
53adfe1c 230     Ok(gix_discover::repository::Path::from_dot_git_dir(
7bed2a96 231         dot_git,
7bed2a96 232         if bare {
20f962e5 233             gix_discover::repository::Kind::PossiblyBare
7bed2a96 234         } else {
53adfe1c 235             gix_discover::repository::Kind::WorkTree { linked_git_dir: None }
7bed2a96 236         },
7d8d1678 237         &gix_fs::current_dir(caps.precompose_unicode)?,
7bed2a96 238     )
7bed2a96 239     .expect("by now the `dot_git` dir is valid as we have accessed it"))
7bed2a96 240 }
7bed2a96 241
0ec2389e 242 fn key(name: &'static str) -> section::ValueName<'static> {
0ec2389e 243     section::ValueName::try_from(name).expect("valid key name")
7bed2a96 244 }
7bed2a96 245
7bed2a96 246 fn bool(v: bool) -> &'static str {
7bed2a96 247     match v {
7bed2a96 248         true => "true",
7bed2a96 249         false => "false",
7bed2a96 250     }
7bed2a96 251 }

gitoxide ( gix-blame) [$] took 8s
❯ sgs
+ 0 gix-blame   # Add initial implementation and tests for `gix-blame`.
+ 1 gix-api     # feat: Add `blame` plumbing crate to the top-level.
+ 2 gix-cli     # feat: add `gix blame` to the CLI
+ 3 review      # first review round
> 4 structure   # modularlize `gix-blame/lib.rs`
- 5 unify-lines # unify how lines in blame results are accessed
- 6 slimdown    # remove duplication and unnecessary parameter
- 7 add-stats   # feat: Add support for statistics and additional performance information.
- 8 error       # add `Error` type
- 9 todos       # Replace todos!() with... something else

gitoxide ( gix-blame) [$]
❯ cargo run --bin gix -- --verbose --trace blame gix/src/create.rs
```

Reply:
Maybe this strikes you as wrong immediately, any help with this is appreciated. In any case, I won't merge until this is sorted out 😅.

Reply:
From what I can tell, this seems to be an issue with `entries_with_lines` and would be solved by using `range_in_blamed_file` instead of `range_in_original_file` (just confirmed locally). It appears I just switched those two names at some point and stopped paying attention. :-)

Reply:
That's it, thanks so much :D! I fell for it, even though I noticed that these are swapped, and even wrote it down somewhere, but didn't fix it.
Will make it a checkbox now :P.

Reply:
You’re welcome! Your review really helped a lot in uncovering these kinds of issues. :-)

Reply:
Ok, it's sorted and I think it works well. All I did was to review the docs and rename `...original...` to `...source...` which is a term I also saw in the Git code.

That makes now sense and it seems to have fallen into place quite nicely.

There is still a couple of things to do, but I think it's Ok to call it a day early due to Christmas and try to wrap it up tomorrow for good.

Try to go back to the third commit, and see how much the code change to something much more readable, it's stunning, and mostly basic refactoring. It's also admirable that `process_change` didn't change at all, as the core is already great and it's really nice to see the drawings, they help tremendously.